ENH: Allow column grouping in DataFrame.plot

NicolasHug opened the pull request:
- [X] closes #29688
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This is an attempt at #29688 i.e. group some columns on the same ax when calling `df.plot`

See [this gist](https://gist.github.com/NicolasHug/4d1f60212ba69e6000982951ae47f203) for illustration of different supported plots

Thanks!

NicolasHug committed:
WIP

NicolasHug committed:
minimal test

WillAyd commented:
@NicolasHug is this still active? Looks like a merge conflict and CI failure need to be addressed

NicolasHug commented:
I'm still happy to work on this. I just haven't received initial feedback yet (and don't want to push either, I understand this isn't necessarily a high priority)

> Before working more on this I'd appreciate some feedback to make sure this is going the in right direction, and that this is actually a desired feature.
Feel free to close otherwise, or LMK if you need to see more to decide :)

WillAyd commented:
Understood. I doubt anyone has reviewed since CI has been red, so definitely want to get that green to start

NicolasHug commented:
Will do. It's just linting though...

NicolasHug committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into subplot_groups

NicolasHug committed:
formatted files

NicolasHug committed:
linting again

WillAyd commented:
@charlesdong1991 any thoughts on this?

charlesdong1991 reviewed (changes requested):
I've noticed this feature request a while ago and personally really like this feature and find it quite potentially useful ^^ @WillAyd  but I think the behaviour of this needs to be well defined.

Many thanks for the PR, it's very nice! @NicolasHug  some minor comments:
1. As mentioned in review, the input for `subplots` needs to be carefully defined, thus only checking if it is `Iterable` or `bool` might not be sufficient.
2. And it would be very nice if you could add some detailed docstring for `subplots` parameter because this is a behaviour change, i think somewhere in `PlotAccessor` in `_core.py`. 
3. And some more tests would be appreciated (see comments below)
4. Nicer to have a docstring for `_col_idx_to_axis_idx` although i think it is quite straightforward
4. Nicer to have a whatsnew note in `1.1.0` rst !

NicolasHug committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into subplot_groups

NicolasHug committed:
Check number of lines and labels

NicolasHug committed:
Added input checks

NicolasHug committed:
some comments

NicolasHug committed:
Whatsnew entry

NicolasHug committed:
More tests

NicolasHug commented:
Thanks a lot for the review @charlesdong1991 , I addressed your comments and added a bunch of validation / tests.

It looks like the code works for most kinds of plots. You can see a few illustrations in [this gist](https://gist.github.com/NicolasHug/4d1f60212ba69e6000982951ae47f203).

Looks like the subplot titles could be improved in some cases, I'll see what I can do. But otherwise I think this is ready for another round ;)

NicolasHug commented:
Regarding the docstring, I'm a bit confused because it seems that `subplots` have been removed from most docstrings? i.e. `git grep "subplots :"` only gives one hit, for `boxplot_frame_groupby()`. I can't find it for `PlotAccessor`.

NicolasHug committed:
Skip test if no scipy for kde

NicolasHug committed:
Merge branch 'master' into subplot_groups

charlesdong1991 reviewed (commented):
very nice! some minor comments

Could you also please make some docs for `subplots`? I think you could add it in `PlotAcessor`, there are some missing docs for parameters, so very nice this one is found out! ^^

NicolasHug committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into subplot_groups

NicolasHug commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -115,6 +116,67 @@ def __init__(
+                )
+
+            if any(
+                not isinstance(group, Iterable) or isinstance(group, str)
```
The thing is, we want `group` to be an iterable but we don't want `group` to be a str, which is also iterable.

Only using `if any(not isinstance(group, Iterable)` would allow strings and we don't want that

NicolasHug commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -115,6 +116,67 @@ def __init__(
+                "kde",
+                "density",
+                "area",
+                "pie",
```
I agree it doesn't look good, but it seems to be consistent (i.e. equally bad) as when `subplots=True`. I updated the [notebook](https://gist.github.com/NicolasHug/4d1f60212ba69e6000982951ae47f203) at the end where you can see how it looks. So maybe it can be kept? No strong opinion!

charlesdong1991 replied:
i have no strong opinion either, my only concern is for pie chart, if they share the same category, they will have the same color in the pie, but different sizes depending on the corresponding values. 

but no strong opinion

NicolasHug committed:
Address comments

NicolasHug committed:
fix docstring

NicolasHug committed:
Added error when columns are fouund in multiple subplots

NicolasHug committed:
Merge branch 'subplot_groups' of github.com:NicolasHug/pandas into subplot_groups

NicolasHug committed:
sorted imports

charlesdong1991 commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -98,6 +98,10 @@ Other enhancements
   (:issue:`33196`)
+- :meth:`DataFrame.plot` will now allow the ``subplots`` parameter to be
+  a list of iterables specifying column groups, so that columns may be
+  grouped together in the same subplot. (:issue:`29944`)
```
we usually put issue number here, not the PR number, so in this case, would use `29688`

charlesdong1991 commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -114,7 +115,75 @@ def __init__(
         self.sort_columns = sort_columns
 
-        self.subplots = subplots
+        if isinstance(subplots, bool):
```
Pandas started to add annotations for the new change a while ago, and now we would prefer to have it annotated especially given this will be a significant change to `subplots`, so could you please add annotations above in init?

charlesdong1991 replied:
you might need to also add some annotations in your code to make mypy happy, if it is the case, feel free to add

charlesdong1991 commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -676,9 +748,21 @@ def _get_ax_layer(cls, ax, primary=True):
         else:
             return getattr(ax, "right_ax", ax)
 
+    def _col_idx_to_axis_idx(self, col_idx):
```
the annotation also is needed for the new method, thanks

charlesdong1991 reviewed (changes requested):
Hey, thanks very much for your work, let's give it another push! 

could you please rebase and resolve conflicts, having some annotations will be good!

jreback commented:
@NicolasHug can you merge master and update to comments

NicolasHug commented:
sorry I'm short on time ATM and I won't be able to come back to this before a few weeks

charlesdong1991 commented:
@nicholsn thanks for this nice proposal, it definitely will be nice to have!!

and I am very sorry for taking so long for the feedbacks!

not sure if I am allowed to finish this PR for you to make sure you have the credit as well.  I will take a look and see!

NicolasHug commented:
no worries at all, I know what it's like ;)

I definitely plan to come back to it later, but if you'd rather have this in sooner then feel free to take over!

charlesdong1991 commented:
ahh, great!! then will leave it on you, and wish to see you coming back to implement this nice feature! ^^  @NicolasHug

NicolasHug committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into subplot_groups

NicolasHug committed:
Added type annotations

NicolasHug commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -120,7 +121,77 @@ def __init__(
+                    "were found in multiple sublots."
+                )
+
+            cols_in_groups_set = set(cols_in_groups)
```
I would rather just do `cols_in_groups = set(cols_in_groups)` but I couldn't make pypy happy otherwise

simonjayhawkins replied:
Again, if you want to reuse variable names for different types, needs a variable annotation before first assignment. I think OK to use different name.

NicolasHug commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -120,7 +121,77 @@ def __init__(
+        if isinstance(subplots, bool):
+            self.subplots = (
+                subplots
+            )  # type: Optional[Union[bool, Sequence[Sequence[Union[str, int]]]]]
```
for some reason I have to declare the type here once again despite having specified it already above.

Also note I have to add `int` so that mypy doesn't complain, but users should only pass strings (column names) instead of ints (indices)

simonjayhawkins replied:
ci failing with Check for use of comment-based annotation syntax. please use py3.6+ syntax https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html?highlight=variable%20annotation#explicit-types-for-variables

> for some reason I have to declare the type here once again despite having specified it already above.

from https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html#type-inference

> Mypy considers the initial assignment as the definition of a variable.

self.subplots is hence infered to be bool without the variable type annotation. (and the mypy error then makes sense `error: Incompatible types in assignment (expression has type "List[<nothing>]", variable has type "bool")`

could use an alias to reduce duplication, and/or could declare variable type in class scope to reduce noise.

also maybe Sequence[Sequence[Union[str, int]]]]] -> Sequence[Sequence[Label]]]]

simonjayhawkins replied:
I think that this is adding quite a bit of code to \_\_init__. maybe create a _validate_subplot_kwarg method (the variable annotation issue could disappear (with function parameters and return type typed)

NicolasHug committed:
some more

NicolasHug committed:
typos

simonjayhawkins commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -84,7 +85,7 @@ def __init__(
         kind=None,
         by=None,
-        subplots=False,
+        subplots: Optional[Union[bool, Sequence[Sequence[Union[str, int]]]]] = False,
```
using `list of iterables` in docstring. mostly we want these consistent.

simonjayhawkins replied:
don't need the Optional

simonjayhawkins commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -120,7 +121,77 @@ def __init__(
+
+            cols_in_groups = [col for group in subplots for col in group]
+            duplicates = {
+                col for (col, cnt) in Counter(cols_in_groups).items() if cnt > 1
```
do we need the counter, can you compare the length of a set with the len of list

NicolasHug replied:
Using a counter allows printing the duplicates. let me know what you prefer

simonjayhawkins replied:
fair enough, if that's a requirement.

NicolasHug committed:
Addressed comments

NicolasHug committed:
Specify stricter output type

NicolasHug committed:
Again

NicolasHug committed:
hopefully fix mypy?

NicolasHug commented:
thanks for the review and for the tips @simonjayhawkins , I have updated the PR accordingly.
The test failure seems to be related to a networking issue (unclosed socket), I'm not sure this is directly related to this PR?

NicolasHug committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into subplot_groups

NicolasHug committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into subplot_groups

NicolasHug commented:
Looks like the CI failure was magically resolved. I think this is ready for another round! thanks for the reviews so far.

NicolasHug committed:
added test for coverage

NicolasHug committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into subplot_groups

jreback reviewed (changes requested):
can you show a sample plot with this api. are there other ways to do this, e. g a for loop?

charlesdong1991 commented:
> can you show a sample plot with this api.

@NicolasHug indeed made a nice sample ^^, see in PR description on top: 
> See this gist for illustration of different supported plots

the link is https://gist.github.com/NicolasHug/4d1f60212ba69e6000982951ae47f203  
there was a discussion about pie plots since they don't as good as others, https://github.com/pandas-dev/pandas/pull/29944#discussion_r404364673 @jreback your opinion is very welcomed!

NicolasHug committed:
changed to whatsnew 112

NicolasHug commented:
Thanks @jreback and @charlesdong1991 . All of your requested changes were addressed.

NicolasHug commented:
> are there other ways to do this, e. g a for loop?

I know of no other way in pandas. Like any other supported plot in pandas, this is of course possible with pure matplotlib code, but with more boilerplate.

NicolasHug commented:
I'm going to stop addressing merge conflicts because they seem to occur 5 minutes after every single commit that I make. They're only about conflicts in the imports. I hope this doesn't hurt reviewing.

charlesdong1991 commented:
thanks @NicolasHug , IMHO, this is a very nice PR and is very close to be merged, many thanks for your patience!

i think the conflict won't affect your PR, so it's okay to resolve the conflict once @jreback comments on https://github.com/pandas-dev/pandas/pull/29944#issuecomment-687559969 and there is kinda consensus on how to address.

One other thing worth mentioning is https://github.com/pandas-dev/pandas/pull/29944#discussion_r483902388 this PR should be merged to  `Plotting` section in `1.2` instead of `1.1.2`, since `1.1.2` will be mainly focused on regression related issues.

NicolasHug committed:
used proper whatsnew file

NicolasHug committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into subplot_groups

NicolasHug committed:
Merge branch 'master' into subplot_groups

NicolasHug commented:
i'm still happy to address any future review

arw2019 commented:
@NicolasHug thanks for being patient! Can you merge master one more time and resolve conflicts?

cc @jreback  xref https://github.com/pandas-dev/pandas/pull/29944#issuecomment-687559969

NicolasHug committed:
Merge branch 'master' into subplot_groups

NicolasHug committed:
unwgnjgrwnjwgnjrkknrjg

arw2019 commented:
cc @jreback @charlesdong1991 this looks close to going in in case we want it for 1.2

arw2019 reviewed (commented):
@NicolasHug sorry about the delay here! Could you merge master again (there are some conflicts) and move the whatsnew to 1.3

arw2019 commented:
ping @jreback @charlesdong1991

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -212,6 +212,7 @@ Other enhancements
 - Added :meth:`~DataFrame.set_flags` for setting table-wide flags on a ``Series`` or ``DataFrame`` (:issue:`28394`)
 - :meth:`DataFrame.applymap` now supports ``na_action`` (:issue:`23803`)
 - :class:`Index` with object dtype supports division and multiplication (:issue:`34160`)
+- :meth:`DataFrame.plot` will now allow the ``subplots`` parameter to be a list of iterables specifying column groups, so that columns may be grouped together in the same subplot (:issue:`29688`).
```
move to 1.3

jreback commented on pandas/plotting/_core.py:
```diff
@@ -632,8 +632,15 @@ class PlotAccessor(PandasObject):
-        Make separate subplots for each column.
+    subplots : bool or list of iterables, default False
+
+        - ``False`` - no subplots will be used
```
add versionupdated 1.3

jreback commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -209,6 +209,79 @@ def __init__(
 
+    def _validate_subplots_kwarg(
+        self, subplots: Union[bool, Sequence[Sequence[str]]]
+    ) -> Union[bool, Sequence[Sequence[int]]]:
```
can you give a doc-string that explains what this is doing

e.g. Parameters / Returns / Raises sections

jreback commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -209,6 +209,79 @@ def __init__(
+            not isinstance(group, Iterable) or isinstance(group, str)
+            for group in subplots
+        ):
+            raise ValueError(
```
test hitting this?

jreback commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -209,6 +209,79 @@ def __init__(
+        )
+        if self._kind not in supported_kinds:
+            raise ValueError(
+                "When subplots is an iterable, kind must be "
```
test hitting this?

jreback commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -209,6 +209,79 @@ def __init__(
+        if isinstance(subplots, bool):
+            return subplots
+        elif not isinstance(subplots, Iterable):
+            raise ValueError("subplots should be a bool or an iterable")
```
test hitting these cases?

NicolasHug replied:
This is already tested

jreback commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -209,6 +209,79 @@ def __init__(
+        duplicates = {col for (col, cnt) in Counter(cols_in_groups).items() if cnt > 1}
+        if duplicates:
+            raise ValueError(
+                f"Each column should be in only one subplot. Columns {duplicates} "
```
test hitting this?

jreback commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -209,6 +209,79 @@ def __init__(
         self._validate_color_args()
 
+    def _validate_subplots_kwarg(
+        self, subplots: Union[bool, Sequence[Sequence[str]]]
```
the return signature is key here Union[bool, List[Tuples[int]]]

jreback reviewed (changes requested):
don't we *already* do this when we `df.groupb(...).plot.line()` for example?

NicolasHug committed:
Merge branch 'master' of github.com:pandas-dev/pandas into subplot_groups

NicolasHug committed:
Addressed comments

pep8speaks commented:
Hello @NicolasHug! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/plotting/_matplotlib/core.py`](https://github.com/pandas-dev/pandas/blob/34515a84f11264a74c22a61e707af05d9ee8862c/pandas/plotting/_matplotlib/core.py):

> [Line 1:15](https://github.com/pandas-dev/pandas/blob/34515a84f11264a74c22a61e707af05d9ee8862c/pandas/plotting/_matplotlib/core.py#L1): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

NicolasHug committed:
remove code due to merge mess up

NicolasHug commented:
Thank you for the review @jreback, comments were addressed.

> don't we already do this when we df.groupb(...).plot.line() for example?

I'm probably missing something but `groupby.plot` does something very different from the proposed PR: `groupby.plot` will group by **value** while the proposed PR simply groups **entire columns** within the same subplots.

NicolasHug commented:
CI Failures are unrelated I think

simonjayhawkins commented:
> CI Failures are unrelated I think

```
Performing static analysis using mypy
pandas/plotting/_matplotlib/core.py:304: error: Incompatible return value type (got "List[Tuple[int, ...]]", expected "Union[bool, List[Tuple[int]]]")  [return-value]
Found 1 error in 1 file (checked 1177 source files)
Performing static analysis using mypy DONE
```

mypy needs to be green before we can consider merging.

NicolasHug committed:
fix mypy

NicolasHug commented:
Thanks, fixed mypy. Should `Linux py38_np_dev` still fail, I believe it's still unrelated to the proposed changes.

NicolasHug committed:
Merge branch 'master' of github.com:pandas-dev/pandas into subplot_groups

NicolasHug committed:
Merge branch 'master' of github.com:pandas-dev/pandas into subplot_groups

NicolasHug committed:
Merge branch 'master' of github.com:pandas-dev/pandas into subplot_groups

lithomas1 commented:
@NicolasHug Mind merging master/resolving conflicts? This would be quite nice to get into 1.3 I think.

NicolasHug committed:
Merge branch 'master' into subplot_groups

NicolasHug committed:
Merge branch 'master' of github.com:pandas-dev/pandas into subplot_groups

NicolasHug committed:
merge again

charlesdong1991 commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -100,7 +105,7 @@ def __init__(
         kind=None,
         by=None,
-        subplots=False,
+        subplots: bool | Sequence[Sequence[str]] = False,
```
I _think_ in general, pandas uses `Union` instead of `|`, would be nicer to align with other similar cases in the codebase.

charlesdong1991 replied:
ah, okay! sorry about that, i was not aware of such update in pre-commit hook.

charlesdong1991 replied:
emm, interesting, still see `Union` used in codebase. maybe @simonjayhawkins or @MarcoGorelli knows better about this change in the precommit hook?

charlesdong1991 reviewed (approved):
Hi, sorry for the wait. I like this PR and the feature, and will be a nice one to have in 1.3! Just a minor comment, otherwise look very good to me.

lithomas1 commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -220,6 +224,98 @@ def __init__(
 
+    def _validate_subplots_kwarg(
+        self, subplots: bool | Sequence[Sequence[str]]
+    ) -> bool | [tuple[int, ...]]:
```
Looks like mypy is failing here.
From https://github.com/pandas-dev/pandas/pull/29944/checks?check_run_id=2548207329
```
pandas/plotting/_matplotlib/core.py:229: error: Bracketed expression "[...]" is not valid as a type  [misc]
pandas/plotting/_matplotlib/core.py:229: note: Did you mean "List[...]"?
```
I think this should be ``list[tuple[int, ...]]``

MarcoGorelli replied:
Hi @NicolasHug - are you sure it was removed by a pre-commit job? I just tried:
```
git checkout -b pr/NicolasHug-subplot_groups master
git fetch https://github.com/NicolasHug/pandas.git subplot_groups
git reset --hard FETCH_HEAD
git checkout 827c5b5
pre-commit run --files pandas/plotting/_matplotlib/core.py
```
and all the checks passed. Indeed, I don't know of any hook which would split an annotation like that in `1517ae3` - perhaps it was an error while fixing a merge conflict? Else, do you happen to remember what you ran? Thanks

NicolasHug committed:
mypy for the billionth time

charlesdong1991 commented on pandas/tests/plotting/frame/test_frame.py:
```diff
@@ -2039,6 +2039,100 @@ def test_plot_no_numeric_data(self):
+        ):
+            df.plot(subplots=[("a", "b")], kind=kind)
+
+    def test_missing_markers_legend(self):
+        # 14958
+        df = DataFrame(np.random.randn(8, 3), columns=["A", "B", "C"])
+        ax = df.plot(y=["A"], marker="x", linestyle="solid")
+        df.plot(y=["B"], marker="o", linestyle="dotted", ax=ax)
+        df.plot(y=["C"], marker="<", linestyle="dotted", ax=ax)
+
+        self._check_legend_labels(ax, labels=["A", "B", "C"])
+        self._check_legend_marker(ax, expected_markers=["x", "o", "<"])
+
+    def test_missing_markers_legend_using_style(self):
+        # 14563
+        df = DataFrame(
+            {
+                "A": [1, 2, 3, 4, 5, 6],
+                "B": [2, 4, 1, 3, 2, 4],
+                "C": [3, 3, 2, 6, 4, 2],
+                "X": [1, 2, 3, 4, 5, 6],
+            }
+        )
+
+        fig, ax = self.plt.subplots()
+        for kind in "ABC":
+            df.plot("X", kind, label=kind, ax=ax, style=".")
+
+        self._check_legend_labels(ax, labels=["A", "B", "C"])
+        self._check_legend_marker(ax, expected_markers=[".", ".", "."])
```
emm, are these 2 tests by mistake added from other PRs?

charlesdong1991 reviewed (approved):
It looks good to me except for a comment about tests from other pr?

In addition to @jreback , maybe @MarcoGorelli or @ivanovmg could also have time to take a look?

ivanovmg commented on pandas/plotting/_core.py:
```diff
@@ -648,8 +648,17 @@ class PlotAccessor(PandasObject):
-    subplots : bool, default False
-        Make separate subplots for each column.
+    subplots : bool or list of iterables, default False
+
```
Just to clarify, are the docs rendered correctly with this empty line?

NicolasHug replied:
We need an empty line because we're starting a list.
Is there a CI job that builds the docs? I couldn't find it

ivanovmg replied:
Yes, CI runs to build the docs.
But you can build the particular page (for ``DataFrame.plot``) yourself to ensure it is correct.
https://pandas.pydata.org/pandas-docs/dev/development/contributing_documentation.html#building-the-documentation

NicolasHug replied:
I couldn't build the docs

```
python make.py clean
python make.py --single pandas.DataFrame.plot
```

fails with stuff like `WARNING: autodoc: failed to import method 'plot.scatter' from module 'DataFrame'; the following exception was raised:
No module named 'DataFrame'`

So instead I just added a line so that the docstring for subplots is consistent with the rest. If the rest is fine, then this one should be fine too.

ivanovmg commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -220,6 +224,98 @@ def __init__(
+        )
+        if self._kind not in supported_kinds:
+            raise ValueError(
+                "When subplots is an iterable, kind must be "
```
What if ``subplots`` is boolean, then would it be necessary to validate ``self._kind``?
If so, then, I guess, it deserves a separate method to be invoked at the moment of ``self.kind`` initialization.

By the way, this is quite confusing for me (probably regardless of the given PR): inside ``MPLPlot`` one initializes ``self.kind``, but it seems that it is not used anywhere, only ``self._kind`` does.
Am I missing something?

NicolasHug replied:
> What if subplots is boolean, then would it be necessary to validate self._kind?

When subplots is boolean we return immediately as we fall back to the current behaviour, as in master. I haven't double checked but I would hope that `self._kind` is validated separately - I mean, it's `_kind`, not `kind`.

The validation happening in `_validate_subplots_kwarg` is only for `subplots`, and its interactions with other parameters (such as `_kind`) when `subplots` is an iterable. But its job is not to validate `_kind`.

NicolasHug committed:
Merge branch 'master' of github.com:pandas-dev/pandas into subplot_groups

ivanovmg commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -736,9 +835,21 @@ def _get_ax_layer(cls, ax, primary=True):
         else:
             return getattr(ax, "right_ax", ax)
 
+    def _col_idx_to_axis_idx(self, col_idx: int):
```
Can you type a return value?

ivanovmg commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -736,9 +835,21 @@ def _get_ax_layer(cls, ax, primary=True):
+            for group_idx, group in enumerate(self.subplots):
+                if col_idx in group:
+                    return group_idx
+        else:
```
Is this ``else`` clause required? I guess, that it may be dropped.

ivanovmg replied:
I suppose, that once you added return type, ``mypy`` started complaining.
Inside the loop you return only in case ``if col_idx in group``, this is the reason for the ``mypy`` error.
I suggest that you modify it:
```
for group_idx, group in enumerate(self.subplots):
    if col_idx in group:
    return group_idx
raise KeyError(f"Failed to find {col_idx} in {self.subplots}")
```
Or something like this.

NicolasHug replied:
I think mypy is wrong here. Because of the validation done earlier, there *must* be a valid column. Raising an exception might make mypy happy but it would never be hit (because of the validation), and we wouldn't be able to test against it. I changed the for loop into a `next()`, which seems to trick mypy into being reasonable again.

ivanovmg commented on pandas/tests/plotting/frame/test_frame.py:
```diff
@@ -2039,6 +2039,72 @@ def test_plot_no_numeric_data(self):
+            (("a",), "each entry should be a list/tuple"),  # iterable of strings
+        ],
+    )
+    def test_group_subplot_bad_input(self, subplots, expected_msg):
```
Good parametrization!

ivanovmg commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -220,6 +224,98 @@ def __init__(
+        duplicates = {col for (col, cnt) in Counter(cols_in_groups).items() if cnt > 1}
+        if duplicates:
+            raise ValueError(
+                f"Each column should be in only one subplot. Columns {duplicates} "
```
Is it really a desirable behavior?
Why not allowing one to plot the same line against different sets?

NicolasHug replied:
Currently (in master), only one column per plot is possible.
Allowing more than one column per plot should be possible, but would make the PR more complex. This PR il already 1.5+ years long so I would leave that improvement for a subsequent PR.

charlesdong1991 replied:
yeah, agree, could be a follow-up enhancement PR!

ivanovmg commented on pandas/tests/plotting/frame/test_frame.py:
```diff
@@ -2039,6 +2039,72 @@ def test_plot_no_numeric_data(self):
+        d = {"a": np.arange(10), "b": np.arange(10), "c": np.arange(10)}
+        df = DataFrame(d)
+
+        with pytest.raises(ValueError, match="should be in only one subplot"):
```
Is it really a desirable behavior (see other comment).

ivanovmg reviewed (commented):
It seems like a reasonable extension of ``subplots`` functionality!
I have one concern regarding the inability to plot the same column on two different subplots (reflected in the comments).
Otherwise looks good!

NicolasHug committed:
add return type

NicolasHug committed:
add line

ivanovmg reviewed (approved):
Looks good to me.
Pending CI checks.

NicolasHug committed:
mypy for the billionth time + 1

jreback commented:
@NicolasHug was looking good. if you want to continue, pls merge master and can relook.

NicolasHug committed:
Merge branch 'master' of github.com:pandas-dev/pandas into subplot_groups

jreback commented:
sorry for the delay here

@NicolasHug was looking good. if you want to continue, pls merge master and can relook.

jreback commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -259,6 +259,7 @@ Other enhancements
 - :meth:`DataFrame.applymap` can now accept kwargs to pass on to the user-provided ``func`` (:issue:`39987`)
 - Passing a :class:`DataFrame` indexer to ``iloc`` is now disallowed for :meth:`Series.__getitem__` and :meth:`DataFrame.__getitem__` (:issue:`39004`)
 - :meth:`Series.apply` can now accept list-like or dictionary-like arguments that aren't lists or dictionaries, e.g. ``ser.apply(np.array(["sum", "mean"]))``, which was already the case for :meth:`DataFrame.apply` (:issue:`39140`)
+- :meth:`DataFrame.plot` will now allow the ``subplots`` parameter to be a list of iterables specifying column groups, so that columns may be grouped together in the same subplot (:issue:`29688`).
```
can you move to 1.5

NicolasHug committed:
Merge branch 'main' of github.com:pandas-dev/pandas into subplot_groups

NicolasHug committed:
update whatnew

NicolasHug commented:
PR is up to date with `main` @jreback

jreback commented:
@NicolasHug can you merge master and we'll see if can get in

NicolasHug committed:
Merge branch 'main' into subplot_groups

NicolasHug committed:
import Iterable from typing.

NicolasHug committed:
Merge branch 'main' of github.com:pandas-dev/pandas into subplot_groups

mroeschke committed:
Merge remote-tracking branch 'upstream/main' into subplot_groups

mroeschke committed:
Process groups to indices in one pass

mroeschke committed:
Merge remote-tracking branch 'upstream/main' into subplot_groups

mroeschke committed:
Fix typing check

mroeschke reviewed (approved):
Thanks for sticking with this @NicolasHug. I merged master and did a small refactor to validate the subplot groups in one pass.

I think this looks pretty good to merge. I believe it requires @jreback approval to move forward.
